### PR TITLE
Prevent slice make and copy in Pushes

### PR DESCRIPTION
**Pull request by JekaMas**
This reduces allocations and CPU

### CODE REVIEW

File: common/big.go [hunk 1]
```diff
@@ -39,3 +39,29 @@ var (
 	Num32 = uint256.NewInt().SetUint64(32)
 	Num35 = uint256.NewInt().SetUint64(35)
 )
+
+func SetBytesRightPadded(z *uint256.Int, buf []byte, length int) *uint256.Int {
```
Comment (AlexeyAkhunov) [inline common/big.go:43]: This is the variation of this, right? https://github.com/holiman/uint256/blob/master/uint256.go#L26

File: core/vm/instructions.go [hunk 2]
```diff
@@ -820,18 +820,17 @@ func makePush(size uint64, pushByteSize int) executionFunc {
...
-		if startMin+pushByteSize < endMin {
-			endMin = startMin + pushByteSize
+		endMin := startMin + pushByteSize
+		if startMin+pushByteSize >= codeLen {
+			endMin = codeLen
 		}
 
 		integer := new(uint256.Int)
-		callContext.stack.push(integer.SetBytes(common.RightPadBytes(callContext.contract.Code[startMin:endMin], pushByteSize)))
+		callContext.stack.push(common.SetBytesRightPadded(integer, callContext.contract.Code[startMin:endMin], pushByteSize))
```
Comment (yperbasis) [inline core/vm/instructions.go:834]: I think we don't need to pad bytes at all: integer.SetBytes does that automatically. The line can be simplified to
callContext.stack.push(integer.SetBytes(callContext.contract.Code[startMin:endMin]))

Comment (AlexeyAkhunov) [reply core/vm/instructions.go:834]: Potentially `common.RightPadBytes` call there was not really required

File: common/big.go (hunk 1 above)
Comment (JekaMas) [reply common/big.go:43]: It is.

File: core/vm/instructions.go (hunk 2 above)
Comment (JekaMas) [reply core/vm/instructions.go:834]: A nice idea. I tried to add this to `TestRightPadding`:
```
    TestRightPadding: big_test.go:28: not equal n=1 l=2 slice=[1]
        Got		[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0]
        !=
        Expect	[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]((*uint256.Int)(0xc0000189c0)((len=4 cap=4) {
         (uint64) 1,
         (uint64) 0,
         (uint64) 0,
         (uint64) 0
        })
        )
```

I think the padding is essential here

Comment (yperbasis) [reply core/vm/instructions.go:834]: Sorry, my bad. We only need to right-pad if endMin-startMin != pushByteSize, which only happens if we're overflowing codeLen.

Comment (JekaMas) [reply core/vm/instructions.go:834]: A good idea, I'll and the check.

Comment (yperbasis) [reply core/vm/instructions.go:834]: But then, on the other hand, your method SetBytesRightPadded has little overhead compared to SetBytes, and we can call it all the time regardless.

Comment (holiman) [reply core/vm/instructions.go:835]: Actually, there's no need to pad, at all. It's possible to cheat here: if we're at the end of the code, the execution will stop. It doesn't matter what's in the stack, so one could just as well just push an empty value.

Comment (yperbasis) [reply core/vm/instructions.go:835]: Thank you Martin for the insight, I've created a PR to simplify: https://github.com/ledgerwatch/turbo-geth/pull/638

**Pull request merged**